coordinator/src/notification_service.rs

```diff
@@ -130,3 +144,205 @@ async fn send_notification<'a>(
+        let position_far_from_expiring_2 = far_from_expiry_position();
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        send_expiry_notifications_if_applicable(
+            &[position_far_from_expiring_1, position_far_from_expiring_2],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_deliving_notifications_before_expiry() {
+        init_tracing_for_test();
+
+        let position_far_from_expiring = far_from_expiry_position();
+        let position_soon_to_expire = soon_expiring_position();
+        let ancient_position = ancient_expired_position(); // too old to send notification
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        send_expiry_notifications_if_applicable(
+            &[
+                position_far_from_expiring,
+                position_soon_to_expire.clone(),
+                ancient_position,
+            ],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 1);
+
+        let notification = received_notifications.first().unwrap();
+        assert_eq!(
+            notification.notification_kind,
+            NotificationKind::PositionSoonToExpire
+        );
+        assert_eq!(notification.user_fcm_token, position_soon_to_expire.1);
+    }
+
+    #[tokio::test]
+    async fn send_only_recently_expired_notifications() {
+        init_tracing_for_test();
+
+        let position_far_from_expiring = far_from_expiry_position();
+        let position_just_expired = just_expired_position();
+        let ancient_position = ancient_expired_position(); // too old to send notification
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        send_expiry_notifications_if_applicable(
+            &[
+                position_far_from_expiring,
+                position_just_expired.clone(),
+                ancient_position,
+            ],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 1);
+
+        let notification = received_notifications.first().unwrap();
+        assert_eq!(
+            notification.notification_kind,
+            NotificationKind::PositionExpired
+        );
+        assert_eq!(notification.user_fcm_token, position_just_expired.1);
+    }
+}
+
+/// Load all recent positions with the DB with associated fcm tokens and send push
+/// notification about expiring/expired positions if needed.
+pub async fn query_and_send_position_notifications(
```

Comment:
🔧 I don't see the notification service doing anything else than to send notifications. This send_expiry_notifications_if_applicable should probably be its own component or maybe added to the expired_positions component?

Reply:
I've renamed the module to `notifications.rs` to avoid confusion. I don't think that a separate module for other free functions would add anything in this case.

it can house the struct and free functions, they're related.